File: backend/app/models/schemas/project.py
"""Project and project-application schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

ProjectAppTerminalType = Literal["web", "h5", "mini_program", "admin", "api", "other"]


class ProjectCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=120)
    team_id: int = Field(..., gt=0)
    product_id: int = Field(..., gt=0)
    description: str | None = None
    is_active: bool = True


class ProjectUpdate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=120)
    team_id: int = Field(..., gt=0)
    product_id: int = Field(..., gt=0)
    description: str | None = None
    is_active: bool = True


class ProjectCopy(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=120)
    is_active: bool = False


class ProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    team_id: int
    team_name: str | None = None
    product_id: int
    product_code: str | None = None
    product_name: str | None = None
    code: str
    name: str
    description: str | None = None
    is_active: bool
    app_count: int = 0
    created_at: datetime
    updated_at: datetime


class ProjectListResponse(BaseModel):
    items: list[ProjectResponse] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 10


class ProjectBulkActionRequest(BaseModel):
    project_ids: list[int] = Field(default_factory=list, min_length=1)
    action: Literal["enable", "disable", "delete"]


class ProjectBulkActionResponse(BaseModel):
    action: Literal["enable", "disable", "delete"]
    affected_ids: list[int] = Field(default_factory=list)
    affected_count: int = 0


class ProjectAppCreate(BaseModel):
    code: str = Field(..., min_length=1, max_length=120)
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = None
    knowledge_base_id: int | None = Field(default=None, gt=0)
    category_id: int | None = Field(default=None, gt=0)
    default_assistant_id: int | None = Field(default=None, gt=0)
    widget_version: str = Field(default="1.0.0", pattern=r"^\d+\.\d+\.\d+$")
    terminal_type: ProjectAppTerminalType = "web"
    is_active: bool = True


class ProjectAppUpdate(BaseModel):
    code: str = Field(..., min_length=1, max_length=120)
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = None
    knowledge_base_id: int | None = Field(default=None, gt=0)
    category_id: int | None = Field(default=None, gt=0)
    default_assistant_id: int | None = Field(default=None, gt=0)
    widget_version: str = Field(default="1.0.0", pattern=r"^\d+\.\d+\.\d+$")
    terminal_type: ProjectAppTerminalType = "web"
    is_active: bool = True


class ProjectAppResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    code: str
    name: str
    description: str | None = None
    knowledge_base_id: int | None = None
    knowledge_base_name: str | None = None
    category_id: int | None = None
    category_name: str | None = None
    default_assistant_id: int | None = None
    default_assistant_name: str | None = None
    widget_version: str
    terminal_type: ProjectAppTerminalType = "web"
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ProjectAppListResponse(BaseModel):
    items: list[ProjectAppResponse] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 10


class ProjectAppBulkActionRequest(BaseModel):
    app_ids: list[int] = Field(default_factory=list, min_length=1)
    action: Literal["enable", "disable", "delete"]


class ProjectAppBulkActionResponse(BaseModel):
    action: Literal["enable", "disable", "delete"]
    affected_ids: list[int] = Field(default_factory=list)
    affected_count: int = 0

File: frontend/src/shared/types/project.ts
export interface ProjectSummary {
  id: number;
  team_id: number;
  team_name: string | null;
  product_id: number;
  product_code: string | null;
  product_name: string | null;
  code: string;
  name: string;
  description: string | null;
  is_active: boolean;
  app_count: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectListResponse {
  items: ProjectSummary[];
  total: number;
  page: number;
  page_size: number;
}

export interface ProjectUpsertPayload {
  team_id: number;
  product_id: number;
  code: string;
  name: string;
  description?: string | null;
  is_active: boolean;
}

export interface ProjectCopyPayload {
  code: string;
  name: string;
  is_active: boolean;
}

export type ProjectBulkAction = "enable" | "disable" | "delete";

export interface ProjectBulkActionResponse {
  action: ProjectBulkAction;
  affected_ids: number[];
  affected_count: number;
}

export interface ProjectAppSummary {
  id: number;
  project_id: number;
  code: string;
  name: string;
  description: string | null;
  terminal_type: ProjectAppTerminalType;
  knowledge_base_id: number | null;
  knowledge_base_name: string | null;
  category_id: number | null;
  category_name: string | null;
  default_assistant_id: number | null;
  default_assistant_name: string | null;
  widget_version: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProjectAppListResponse {
  items: ProjectAppSummary[];
  total: number;
  page: number;
  page_size: number;
}

export interface ProjectAppUpsertPayload {
  code: string;
  name: string;
  description?: string | null;
  terminal_type: ProjectAppTerminalType;
  knowledge_base_id?: number | null;
  category_id?: number | null;
  default_assistant_id?: number | null;
  widget_version: string;
  is_active: boolean;
}

export type ProjectAppTerminalType = "web" | "h5" | "mini_program" | "admin" | "api" | "other";

export const PROJECT_APP_TERMINAL_TYPE_LABELS: Record<ProjectAppTerminalType, string> = {
  web: "Web",
  h5: "H5",
  mini_program: "小程序",
  admin: "管理后台",
  api: "API",
  other: "其他",
};

export type ProjectAppBulkAction = "enable" | "disable" | "delete";

export interface ProjectAppBulkActionResponse {
  action: ProjectAppBulkAction;
  affected_ids: number[];
  affected_count: number;
}

export interface ProjectAppEmbedPreviewResponse {
  embed_url: string;
  expires_in_seconds: number;
}

export interface ProjectAppEmbedPreviewPayload {
  store_id?: string | null;
}
